### DISCUSSION TITLE
Ante, a systems game about technological progress

### DISCUSSION THREAD

**Discussion by TheNeikos** [bevyengine/bevy#1474] [Devlogs]
Hello and welcome to my first devlog about Ante!

Ante will be a game about utilizing the various systems and resources existing in the world to achieve your goal of building up the first habitat on a foreign planet. 

It is meant to be technological progression game, with a focus on optimizing resource pipelines as well as efficiently using the tools at your disposal to achieve ever more complex structures.

But, before all that becomes reality the whole things needs to be made, and this is what these devlogs will be all about.

I have already made a very rough prototype ([which you can check out here](https://discord.com/channels/691052431525675048/692648638823923732/808978274210414644)) where I tried to implement all the base features I'd need to get started: Semi-efficient voxel rendering, as well as interacting with said world. The issue is, the computation and the rendering were so far intermingled. Since I think seperating these concerns is beneficial, (and envision of having some networking at some point) I decided to bite the bullet now rather than later and refactor my code.

Previously the setup was rather simple:

1. Bevy startup
2. Generate world data
3. Spawn player
4. Generate mesh data around the player

Given we're using an ECS these steps were all loosely tied together, which allowed me to advance rather quickly. For example: the world is split up into chunks to make things like rendering a bit easier. Then, whenever the player moves around, it's "current chunk" is updated, which in turn instructs another system to check if all chunks exist/are visible around them (commonly called 'render distance'). It then generates the chunk entities with their mesh as needed and spawns it into the world. Chunks that are too far away are removed so as to take off load of the GPU. And any chunks which had their mesh modified will be updated (through their `Handle<Mesh>`)

The new setup is similar, but instead of just a single application it's split into a one-to-many architecture. 
- The server, will be the authoritative source of truth, and accept messages from the outside. Without networking it will be simple channels, coming from the client. In the same vein, it will be able to send data the client requests, (like "Hey, I wanna know how things look like around me"). Since this all happens in the same binary there should be no delay other than the overhead induced from the messaging system, which I expect to be negligible. This system should then also allow me to later incorporate networking if I want to.
- The client, where the world actually gets rendered as well as input from the player is processed to be then sent to the server.

Current line count:
```
> tokei
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 TOML                    4           55           41            4           10
-------------------------------------------------------------------------------
 Rust                   10         1196         1013           42          141
 |- Markdown             1            4            0            3            1
 (Total)                           1200         1013           45          142
===============================================================================
 Total                  14         1251         1054           46          151
===============================================================================
```

**Comment by tigregalis** [comment]
Nice! What would you say is the closest analogue in terms of gameplay? Factorio?

I'd like to know more about the server-client stuff if you don't mind. Or even if you can point me to resources to learn about it. I haven't been able to find anything that explains things in simple terms but also in practical terms (as in guides you towards an actual implementation).

**Reply by TheNeikos** [reply]
> Nice! What would you say is the closest analogue in terms of gameplay? Factorio?

I think you could say Factorio/Satisfactory. But I am not planning on explicitely copying their game loop.

> I'd like to know more about the server-client stuff if you don't mind.

I'm gonna post a follow up soon-ish! And write down what worked so that others may do the same.
